### ISSUE TITLE
too much recursion error on Firefox

### ISSUE THREAD

**Issue by dohomi** [meteor/meteor#3028]
Hello,

I curious if the Meteor team is aware of the too much recursion bug in Firefox:
http://autoform.meteor.com/qfdetails
The error mainly occurs because of the usage of https://docs.meteor.com/#/full/template_parentdata ```Template.parentData```. We receive this error on all older Computer/Mac/Linux, the page is not opening at all. On newer Macs the website opens with some minor errors.

Thanks, Dominic

**Comment by brylie** [comment]
For easy reproduction, I have created a [minimal example using aldeed:autoform](https://github.com/brylie/meteor-quickform-recurstion):

```html
<body>
 {{> quickForm id="quickform" schema="schema"}}
</body>
```

```javascript
schema = new SimpleSchema({
    name: {
        type: String
    }
});
```

**Comment by avital** [comment]
Does someone have a simple repro without external packages? That will make it easier to investigate.

**Comment by lalomartins** [comment]
@avital I was trying to write one but a deeper look at  aldeed/meteor-autoform#333 points to this being not really a Meteor bug but rather a question of Firefox having a smaller recursion limit (at least on Linux). I'd wait and see what they end up with as a fix.

**Comment by darkship** [comment]
I just run into this problem. I'm using linux mint 
I don't use <code>aldeed/meteor-autoform</code> but I have this error on FF 35 and FF 36 both on linux and it doesn't always happen.

I tried http://adamrich.name/recursion.html on different browsers and here are my results:
- Windows 8 firefox 34 : 7k5 to 12k recursions with a peak at 20k
- Windows 8 firefox 35 : 8k to 11k recursions with a peak at 18k
- Windows 8 firefox 36 (dev edition)  : 10k to 13k recursions with a peak at 25k
- Windows 8 chrome 39 : 22k to 24k recursions
- linux mint 17  firefox 35  : 5k to 6k (sometimes 9K) recursions with a peak at 16k
- linux mint 17  firefox 36 (dev edition)  : 5k to 6k recursions with a peak at 16k
- linux mint 17 chromium 39 : 23k recursions

I thought first that was a bug in Firefox dev edition. I just update to  firefox 35 so I have no tests with FF34. 

I have something like 15 nested templates in my app and when I remove 1 or 2 of them, the bug  doesn't show anymore. 

I will add a repo.

**Comment by brylie** [comment]
I ran the browser recursion test too:

```
Stopped at iteration 7181 with error message: too much recursion
```

Thanks for the suggestion :-)

**Comment by brylie** [comment]
That is an interesting recursion repo @darkship. How can I track down the recursion in Meteor templates? I would like to help shine some light on this issue, and hopefully improve Meteor's scalability, etc :-)

**Comment by darkship** [comment]
I noticed that if you replace <code>#with</code> by <code>#if</code> in my repo : 
```
<template name="list">
    {{#if item}}
        {{>item this}}
    {{/if}}
</template>
```
You get an error even if your data has only 1 recursion and you use Chromium.

Going through Blaze, the last piece of code I found that provoke a recursion was <code>return this.visitObject.apply(this, arguments);</code> in  meteor/packages/htmljs/visitors.js line 61.

@brylie I cloned Meteor from Git and and followed the errors

**Comment by benjamn** [comment]
After some investigation, it appears that a combination of deeply nested template contexts in the quickform package, together with relatively deep call stacks in the Blaze implementation, are conspiring to trigger the `RangeError`. In other words, the call stacks in question are large, but not infinitely large, so there is probably no single bug here that we can fix to make the problem go away entirely, but many opportunities to alleviate the problem in realistic use cases.

Two opportunities that we've thought about so far:
* We can make individual stack frames smaller by moving parameters and other local variables onto the heap. Apparently the number of local variables in a function affects the memory footprint of its stack frames, which limits the total number of stack frames that can be nested. So replacing multiple local variables with a single object whose properties represent the original variables might let us get away with more recursion.
* We can reimplement some algorithms that are currently recursive as iterative algorithms, for example `HTML.visitor` (in the htmljs package), which shows up on call stacks for this reproduction.

Given the exploratory nature of these solutions, I'm not sure when we'll be able to provide real results, but we're very glad to know about the problem, and we acknowledge that fixing it will require optimizations in the Blaze implementation.

**Comment by Levino** [comment]
Would love a fix too. We have the issue on coyno.com.

**Comment by laurentpayot** [comment]
Same autoform issue here with Firefox 35.0.1 and Meteor 1.0.3.1 on Ubuntu 14.10. Works fine with Chromium.

**Comment by aldeed** [comment]
An update from the autoform side: I've done some work in `devel` branch to remove unnecessary layers of templates. It's helped a bit but it's still pretty easy to reproduce this issue. I think the primary culprit is the way autoform is using parentData. We recurse through a bunch of layers of templates and then look up the tree with parentData. I don't think parentData itself uses much recursion, but it might be resulting in a reactive dependency that leads to recursion upon reruns.

Still pretty much guessing. Hopefully I can figure out a clean reproduction.

**Comment by brylie** [comment]
Thanks for your work on this issue Eric. :-)

It is also worth noting that the Blaze developers are working on the issue as well. How can we coordinate these efforts?

**Comment by brylie** [comment]
Somewhat off-topic, but interestingly relevant is the paper [Natural language – no infinity
and probably no recursion](http://mindmodeling.org/cogsci2012/papers/0340/paper0340.pdf).

**Comment by BradRyan** [comment]
I'm getting "Maximum call stack size exceeded" in Chrome when Autovalue is taking another field value, this.siblingField().  I am able to store the field value in a local variable and return a value from the nested array even.  The error occurs after I return the value.  

My form is somewhat complex (arrays of objects in a few fields).  Might have to manually create input fields for my use case since this sounds like a complicated fix.  I'm still happy all my other simpler forms are working well!

**Comment by mxab** [comment]
Seems to be worse with the latest Firefox (36) version on Windows 8

**Comment by mxab** [comment]
I made a nested template example which fails in Firefox in Windows 8 (not tested in 7) 
it fails as soon if there are more than 23 nested templates, in chrome it still works if you render more than 150 nested templates
see this example:
https://github.com/mxab/meteor-call-stack-exceed

**Comment by dohomi** [comment]
Hi @benjamn 
Does anybody use Meteor for a production app, because in my opinion this is a major blocker and Blaze shouldn't call itself stable. I didn't recognize this issue, but Meteor is calling itself stable for almost 4 months and the biggest browser Firefox isn't supported yet - how can production apps survive without the 20-40% of visitors who won't be able to use the app?
Would be great to hear any news on this issue. Thanks

**Comment by lalomartins** [comment]
@dohomi it's quite stable; this bug is a corner case that doesn't come up in normal circumstances. You might flip it around and ask if Firefox is stable! (After all it is, technically speaking, a Firefox bug.)

**Comment by dohomi** [comment]
@lalomartins sorry to say but Firefox is one of the major browser at the moment.. Meteor has to adopt its code to Firefox not other way around.. Who can waive all Firefox users on Windows, Linux or older Macs away because you would state its a Firefox bug? Check @mxab simple reproduction package where you see how easy it is to crash your app on Firefox. In the end as a startup you depend on every single visitor and they won't care who's responsible for the black page they are visiting - they simply never come back.

**Comment by lalomartins** [comment]
@dohomi you're entirely missing the point, so let me spell it out for you.

Coming here to disparage the project with “can you really call yourself stable” FUD is unproductive and not acceptable behaviour in an issue tracker.

Your users are correct if they don't come back, because as a startup neither meteor or firefox is responsible for the black page your visitors see — you are. You're the only one here trying to shift blame and responsibility. As a startup it's your job to test thoroughly, and honestly, working around browser and/or framework bugs is at least 30% of a web developer's job description. It's especially unpleasant to see you disparaging a project which, I'm sure, already saved you weeks, if not months, of development time, compared to developing whatever you have on any other platform.

So really, play nice. This is not the space for this kind of talk. In fact, I don't think anywhere is.

**Comment by brylie** [comment]
Hey all. Please consider continuing this discussion on the Meteor forums. At this point it is neither constructive nor salient.

On 1 March 2015 14:01:06 EET, Lalo Martins <notifications@github.com> wrote:
>@dohomi you're entirely missing the point, so let me spell it out for
>you.
>
>Coming here to disparage the project with “can you really call yourself
>stable” FUD is unproductive and not acceptable behaviour in an issue
>tracker.
>
>Your users are correct if they don't come back, because as a startup
>neither meteor or firefox is responsible for the black page your
>visitors see — you are. You're the only one here trying to shift blame
>and responsibility. As a startup it's your job to test thoroughly, and
>honestly, working around browser and/or framework bugs is at least 30%
>of a web developer's job description. It's especially unpleasant to see
>you disparaging a project which, I'm sure, already saved you weeks, if
>not months, of development time, compared to developing whatever you
>have on any other platform.
>
>So really, play nice. This is not the space for this kind of talk. In
>fact, I don't think anywhere is.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/meteor/meteor/issues/3028#issuecomment-76593778

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mxab** [comment]
@lalomartins I would not specify the error as a bug of firefox since I am not aware of any html/js specification about max callstack / recursion limitations ( but also I haven't googled for it hardcore)

I would neither call it a meteor "bug" if you do not take any recursion limitations into account it works as expected, further I think this issue has not surfaced more as there is not official windows development support

due to the tree structured nature of my mini cms I cannot avoid deep nesting of templates

as soon have you have some more "complex" nesting this problem occurs faster
https://github.com/mxab/meteor-call-stack-exceed/tree/more-samples

so I think as @benjamn said, blaze/htmljs needs to be tuned. I tried some peeking in blaze / htmljs parts and was printing out the stack size but I haven't found the point where it "explodes". So if someone could point me in the right direction I really would like to find some workarrounds for this issue

**Comment by mxab** [comment]
Hi @stbaer and did some more research on the blaze package, after inserting the "catch" statement to log the errors at
https://github.com/meteor/meteor/compare/master...stbaer:ff_recursion_testing#diff-61be8b0b328aaf8a6ae2f89b67662389R416
and 
https://github.com/meteor/meteor/compare/master...stbaer:ff_recursion_testing#diff-52584623fe717d42fab5db3d19d9a498R481
we noticed several accesses on undefineds, after preventing theses errors at 

- https://github.com/meteor/meteor/compare/master...stbaer:ff_recursion_testing#diff-c2136e0b7486905f58b92d0c18115363R151
- https://github.com/meteor/meteor/compare/master...stbaer:ff_recursion_testing#diff-e964c030180e303be57f63585acebf0cR90
- https://github.com/meteor/meteor/compare/master...stbaer:ff_recursion_testing#diff-e964c030180e303be57f63585acebf0cR305

The original example allow the Win8/Firefox 36 to easily nest more then 300 templates
see https://github.com/mxab/meteor-call-stack-exceed/tree/with-undefined-catches

@benjamn could this be a possible optimisation or are the undefined access in these cases expected or should be handled differently
I haven't tested it against any other packages like @aldeed autoform

**Comment by aldeed** [comment]
@mxab, sounds promising! I will try to find some time to test those changes against autoform recursion errors.

**Comment by stbaer** [comment]
Hi @aldeed, like @mxab said, the changes work very well for the https://github.com/mxab/meteor-call-stack-exceed example. But in our project with autoform included firefox shows an 'unresponsive script' popup.

@avital or anybody from the meteor team, any thoughts on this?

**Comment by shtefcs** [comment]
To much people have problem with it https://github.com/aldeed/meteor-autoform/issues/333, can you **@MDG** give us some updates on it plsss ?

**Comment by lalomartins** [comment]
Here's the thing though. When I say it's a Firefox issue, I'm not arguing about blame assignment, because that doesn't help solve the problem. What I'm saying is, it's not clear there's anything autoform or meteor can do to fix it. That doesn't mean it's not worth trying, but ultimately, the best fix just might end up being a pull request to Mozilla (or whatever the Mercurial equivalent is).

tl;dr chill

**Comment by dgreensp** [comment]
Fixed on `devel`. Thanks for your patience, everyone.

**Comment by mquandalle** [comment]
wow!

**Comment by benjamn** [comment]
Awesome!

**Comment by ripdog** [comment]
Yes, thank you! Now Firefox can once again be a first class citizen of
my app! :)


On Thu, Mar 26, 2015, at 03:55 PM, Ben Newman wrote:
> Awesome!


> — Reply to this email directly or view it on GitHub[1].

>


Links:

  1. https://github.com/meteor/meteor/issues/3028#issuecomment-86311901

**Comment by boustanihani** [comment]
COOL :)

**Comment by aldeed** [comment]
@dgreensp, hoping you can sneak this into 1.1 or a patch soon after.

**Comment by avital** [comment]
Yup, this is in 1.1-rc.1

**Comment by aldeed** [comment]
Seems to have fixed the autoform issues. Thanks!

**Comment by shtefcs** [comment]
Awesome, tnx guys!

**Comment by dohomi** [comment]
Thanks guys!

**Comment by darkship** [comment]
This is definitely much, much better, thanks !

I updated my repo https://github.com/darkship/test_much_recursion. I could go from 1 to 457 (instead of 1 to 12) recursions. Some errors remains when I go back to 1 but it doesn't seem to be bothering.
